File: SVM/src/main/java/SVMDataGen.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author minli
 * unused
 */
package SVM.src.main.java;
import org.apache.log4j.Logger;
import org.apache.log4j.Level;
import org.apache.spark.mllib.util.SVMDataGenerator;
import org.apache.spark.mllib.regression.LabeledPoint;
import org.apache.spark.api.java.*;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.rdd.RDD;
import org.apache.log4j.Logger;
import org.apache.log4j.Level;

public class SVMDataGen {
            public static void main(String[] args) {
        if (args.length < 5) {
            System.out.println("usage: <output> <nExamples> <nFeatures> <eps> <intercept> [numPar]");
            System.exit(0);
        }
						Logger.getLogger("org.apache.spark").setLevel(Level.WARN);
				Logger.getLogger("org.eclipse.jetty.server").setLevel(Level.OFF);
        String output = args[0];
        int nExamples = Integer.parseInt(args[1]);
        int nFeatures = Integer.parseInt(args[2]);
        double eps = Double.parseDouble(args[3]);
        double intercepts = Double.parseDouble(args[4]);
        int numPar = (args.length > 5) ? Integer.parseInt(args[5]) : System.getProperty("spark.default.parallelism") != null ? Integer.parseInt(System.getProperty("spark.default.parallelism")) : 2;
        
        SparkConf conf = new SparkConf().setAppName("SVM data generation");
        //JavaSparkContext sc = new JavaSparkContext(conf);
        SparkContext sc = new SparkContext(conf);
        //RDD<double[]> data = KMeansDataGenerator.generateKMeansRDD(sc, numPoint, numCluster, numDim, scaling, numPar);
        /*RDD<LabeledPoint> data=SVMDataGenerator.generateLinearRDD(sc,nExamples,
                nFeatures,eps,numPar,intercepts);
        JavaRDD<LabeledPoint> tmpdata=data.toJavaRDD();
        JavaRDD<String> parsedData = tmpdata.map(
               new Function<LabeledPoint, String>() {
                    public String call(LabeledPoint s) {                        
                        return s.toString();
                    }
                }
        );
        parsedData.saveAsTextFile(output);*/
    }
}

File: SVM/src/main/scala/SVMDataGen.scala

/*
 * (C) Copyright IBM Corp. 2015 
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at 
 *
 *  http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package SVM.src.main.scala

import org.apache.log4j.Logger
import org.apache.log4j.Level
import scala.util.Random

import com.github.fommil.netlib.BLAS.{getInstance => blas}

import org.apache.spark.SparkContext
import org.apache.spark.annotation.{DeveloperApi, Since}
import org.apache.spark.mllib.linalg.Vectors
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.rdd.RDD

object SVMDataGen {

        def main(args: Array[String]) {
            if (args.length < 2) {
                println("Usage: SVMGenerator " +
                        "<master> <output_dir> [num_examples] [num_features] [num_partitions]")
                    System.exit(1)
            }
	Logger.getLogger("org.apache.spark").setLevel(Level.WARN)
    Logger.getLogger("org.eclipse.jetty.server").setLevel(Level.OFF)

            val sparkMaster: String = args(0)
                val outputPath: String = args(1)
                val nexamples: Int = if (args.length > 2) args(2).toInt else 1000
                val nfeatures: Int = if (args.length > 3) args(3).toInt else 2
                val parts: Int = if (args.length > 4) args(4).toInt else 2

                val sc = new SparkContext(sparkMaster, "SVMGenerator")

                val globalRnd = new Random(94720)
                //val trueWeights = Array.fill[Double](nfeatures + 1)(globalRnd.nextGaussian())
                val trueWeights = Array.fill[Double](nfeatures )(globalRnd.nextGaussian())

                val data: RDD[LabeledPoint] = sc.parallelize(0 until nexamples, parts).map { idx =>
                    val rnd = new Random(42 + idx)

                        val x = Array.fill[Double](nfeatures) {
                            rnd.nextDouble() * 2.0 - 1.0
                        }
                    val yD = blas.ddot(trueWeights.length, x, 1, trueWeights, 1) + rnd.nextGaussian() * 0.1
                        val y = if (yD < 0) 0.0 else 1.0
                        LabeledPoint(y, Vectors.dense(x))
                }

            data.saveAsTextFile(outputPath)

                sc.stop()
        }
}
